pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,498 @@
+		let mut votes = 0;
+
+		for router_id in &inbound_processing_info.router_ids {
+			weight.saturating_accrue(T::DbWeight::get().reads(1));
+
+			match PendingInboundEntries::<T>::get(
+				inbound_processing_info.current_session_id,
+				(message_proof, router_id),
+			) {
+				// We expected one InboundEntry for each router, if that's not the case,
+				// we can return.
+				None => return Ok(()),
+				Some(inbound_entry) => match inbound_entry {
+					InboundEntry::Message {
+						message: stored_message,
+						..
+					} => message = Some(stored_message),
+					InboundEntry::Proof { current_count } => {
+						if current_count > 0 {
+							votes += 1;
+						}
+					}
+				},
+			};
+		}
+
+		if votes < inbound_processing_info.expected_proof_count_per_message {
+			return Ok(());
+		}
+
+		match message {
+			Some(msg) => {
+				Self::decrease_pending_entries_counts(
+					&inbound_processing_info,
+					message_proof,
+					weight,
+				)?;
+
+				T::InboundMessageHandler::handle(
+					inbound_processing_info.domain_address.clone(),
+					msg,
+				)
+			}
+			None => Ok(()),
+		}
+	}
+
+	/// Decreases the counts for inbound entries and removes them if the
+	/// counts reach 0.
+	fn decrease_pending_entries_counts(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message_proof: Proof,
+		weight: &mut Weight,
+	) -> DispatchResult {
+		for router_id in &inbound_processing_info.router_ids {
+			weight.saturating_accrue(T::DbWeight::get().writes(1));
+
+			match PendingInboundEntries::<T>::try_mutate(
+				inbound_processing_info.current_session_id,
+				(message_proof, router_id),
+				|storage_entry| match storage_entry {
+					None => Err(Error::<T>::PendingInboundEntryNotFound.into()),
+					Some(stored_inbound_entry) => match stored_inbound_entry {
+						InboundEntry::Message {
+							expected_proof_count,
+							..
+						} => {
+							let updated_count = (*expected_proof_count).ensure_sub(
+								inbound_processing_info.expected_proof_count_per_message,
+							)?;
+
+							if updated_count == 0 {
+								*storage_entry = None;
+							} else {
+								*expected_proof_count = updated_count;
+							}
+
+							Ok::<(), DispatchError>(())
+						}
+						InboundEntry::Proof { current_count } => {
+							let updated_count = (*current_count).ensure_sub(1)?;
+
+							if updated_count == 0 {
+								*storage_entry = None;
+							} else {
+								*current_count = updated_count;
+							}
+
+							Ok::<(), DispatchError>(())
+						}
+					},
+				},
+			) {
+				Ok(()) => {}
+				Err(e) => return Err(e),
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Retrieves the information required for processing an inbound
+	/// message.
+	pub(crate) fn get_inbound_processing_info(
+		domain_address: DomainAddress,
+		weight: &mut Weight,
+	) -> Result<InboundProcessingInfo<T>, DispatchError> {
+		let router_ids = Self::get_router_ids_for_domain(domain_address.domain())?;
+
+		weight.saturating_accrue(T::DbWeight::get().reads(1));
```

Comment:
To make your life easier, because we're using defensive weights, we do not need to take into account these computations, just returning `LP_DEFENSIVE_WEIGHT` for any processed (inbound or outbound) message is enough

Reply:
Hmm, gotta double-check this since some of these weights are actually returned when the message is handled in the message queue.

Reply:
Oh right, according to [this](https://github.com/centrifuge/centrifuge-chain/blob/feat/lp-v2-gateway-multi-router/pallets/liquidity-pools-gateway/src/message_processing.rs#L405), we can definitely exceed the max processing weight defined [here](https://github.com/centrifuge/centrifuge-chain/blob/88a902d4d2fa5b38ae704f14d5eb5117e7edf444/pallets/liquidity-pools-gateway/src/lib.rs#L540).

So should I go ahead and just remove these small weight accruals for reads/writes?

Reply:
> So should I go ahead and just remove these small weight accruals for reads/writes?

I would do to leave things cleaner to the auditers

Reply:
Ideally, this will be benchmarked so we will end up not having weights mixed with the production code

Reply:
I think leaving the same implementation as it is now in `main` is enough

Reply:
> I think leaving the same implementation as it is now in `main` is enough

I agree!

Reply:
Yup, will change these soon.

Reply:
Done.